### PR TITLE
First fixes and minor tweaks.

### PR DESCRIPTION
**Pull request by patricknelson**
Check this out, here's my stab at it. Basically I fixed a few minor redundancies, unused vars, tried newer approaches at things (since we're requiring modern [read: non-IE] browsers) and using HTML5, so we're free to use those dandy global ID's.

Also, biggest change (so far) was the revision of the `promise.then(...)` in the main `srs.load()` method. That's the resolution of a promise, so that closure (in as the `then()` param) is being passed the data processed by GPG, so I removed that extra "reject" parameter (since that's what you put into a `Promise` constructor anyway) and renamed it a bit. Also, _more importantly_,  I had no way to handle errors, so I used promises as a return type for `srs.load(...)` as well so that, if I wanted to, I could handle them. Otherwise, they will propagate up as `Error`'s anyway :+1: 

I might make more tweaks as I look at it but this is what I've got so far, so here's a PR!

### CODE REVIEW

**Comment by patricknelson** [comment]
Oh! I forgot to mention one of the most important security improvements. I changed this:

``` js
var valid = resolve.signatures.filter(sig => sig.valid).length > 0;
```

to

``` js
// Consider the entire response invalid if any of the signatures are not valid.
var valid = data.signatures.filter(sig => sig.valid).length == data.signatures.length;
```

I figured that made the most sense, since apparently GPG will return multiple possible signatures and, if we're going to be filtering out and counting them anyway, we should be buzzing the alarms if _ANY_ of them aren't valid, **not** if _at least one_ is valid.

**Comment by patricknelson** [comment]
We may also want to consider throwing in maybe a gulp/grunt workflow (for shits/giggles) to transpile to Babel for the less fortunate (IE and beyond). But that's probably way out of scope :smile:

**Comment by patricknelson** [comment]
Another change is that I tweaked the format of the public key supplied inline (even though I know that's not important for final implementation) to follow the more modern standard seen by JavaScript templating where they'll typically encapsulate hidden templates (or text data strictly for JS use) in `<script>` tags.  [Like mustache for example](https://github.com/janl/mustache.js).

'stache

File: README.md [hunk 1]
```diff
@@ -24,6 +24,8 @@ Instead, I propose that developers cryptographically sign the resources they int
 
 The prototype consists of two signed JavaScript files. One has been modified after it was signed. Only the signed file should execute in the browser. The prototype uses [OpenPGPJs](http://openpgpjs.org), and requires a modern browser. Just view tests/index.html and view the browser console.
 
+**Please note:** In practice, embedding the public key anywhere on the site where the content is also hosted (like in this example code) would be effectively useless. To help mitigate risk, it should instead be kept in a centralized and publicly accessible key store, or preferably, using a public key already retained or cached on the client's computer. That centralized mechanism isn't described here. This is only a *shim* which can be implemented in-browser in order to prevent further tampering by a potential hack on the server-side or MITM attack for plain text connections.
+
```
Comment (kojiromike) [inline README.md:28]: This isn't really right. It comes down to a matter of trust. In this case (as with SRI) the user has implicitly trusted the primary server (the one that served the html page with subresources on it). Any embedded JavaScript will be executed. Therefore any embedded public keys can be trusted.

File: src/subresource-signing.js [hunk 2]
```diff
@@ -8,30 +8,45 @@ var srs = {
...
+                        if (!valid) {
+                            var message = 'The file at "' + url + '" does not have a valid signature.';
+                            message += ' Refusing to load that file!';
+                            reject(message);
+                        } else {
+                            srs._.attach_script(data.text);
+                            resolve({
+                                url,
+                                gpg: data
+                            });
```
Comment (kojiromike) [inline src/subresource-signing.js:41]: Nice. Thanks!

File: src/subresource-signing.js [hunk 3]
```diff
@@ -8,30 +8,45 @@ var srs = {
...
+            request.onreadystatechange = function() {
+                if (this.readyState !== 4) return;
+                if (this.status !== 200) {
+                    reject(srs._.http_error(this.status));
+                } else {
+                    var promise = srs._.validate(this.response, armored_pubkey);
+                    promise.then(function(data) {
+                        // Consider the entire response invalid if any of the signatures are not valid.
+                        var valid = data.signatures.filter(sig => sig.valid).length == data.signatures.length;
+                        if (!valid) {
```
Comment (kojiromike) [inline src/subresource-signing.js:32]: Say we have a list of trusted pubkeys. Pubkeys are identities. So for example, I could say I trust the jQuery team's pubkey and my own pubkey, but the jQuery file would be signed by that team and my js would be signed by me. Using this validation, neither file would be allowed.

File: src/subresource-signing.js [hunk 4]
```diff
@@ -39,8 +54,10 @@ var srs = {
         /**
          * Assume the contents is valid, and attach the script
          * to the document so it runs.
+         *
+         * @param contents
          */
-        attach_script: function srs_attach_script(contents) {
+        attach_script: function(contents) {
```
Comment (kojiromike) [inline src/subresource-signing.js:60]: Why did you remove all the function names?

File: tests/index.html [hunk 5]
```diff
@@ -6,48 +6,60 @@
...
+                    console.log(data.url, data.gpg)
+                }
+
+                function whoops(err) {
+                    console.error("Error validating signed resorce: " + err)
+                }
+
+                srs.load(pubkey, 'valid-sig-test.js.asc').then(yay).catch(whoops);
+                srs.load(pubkey, 'invalid-sig-test.js.asc').then(yay).catch(whoops);
+            }(pubkey.innerHTML.replace(/\n +/g, '\n')));
```
Comment (kojiromike) [inline tests/index.html:62]: Just out of curiosity, did you try it without the `replace`? I wonder if openpgpjs's armor parser can handle cleaning the whitespace.

File: src/subresource-signing.js (hunk 3 above)
Comment (patricknelson) [inline src/subresource-signing.js:32]: Makes sense. I didn't realize that's how it worked.

File: src/subresource-signing.js (hunk 4 above)
Comment (patricknelson) [inline src/subresource-signing.js:60]: They're not useful since they only work inside the function in this case or as a quirk in certain versions of IE and we've already sworn them off with all this ES6 fanciness.

File: tests/index.html (hunk 5 above)
Comment (patricknelson) [inline tests/index.html:62]: I assumed that too, because it's standard for `base64` as well, but for some reason, it failed for me. As a last resort (to support my snazzy syntax) I had to write a little more ugly code to compensate :smile: If you can make it work, though, that'd be cool because I like this template format.

File: README.md (hunk 1 above)
Comment (patricknelson) [inline README.md:28]: I don't necessarily agree. The signature can coexist with the code, which is fine, because it's there as a sort of integrity check (similar to the original SRI that this is conceptualized from). Basically what @scottsb said here: https://twitter.com/thescottsb/status/676519476154793984 - The idea here is if the file contents were compromised (hence the need to cryptographically verify contents with pub/priv key) then so could the public key as well, if they live in the same place. So the idea here is to distribute the  potential for compromise for which this is originally intended. Also public key servers are already a common/old concept anyway, especially for PGP/GPG.

File: src/subresource-signing.js (hunk 3 above)
Comment (patricknelson) [inline src/subresource-signing.js:32]: Ok, basically reverted this: https://github.com/patricknelson/Subresource-Signing/commit/f6e5e4152a5acc9462ae56660713cdfca8cdf16f

File: src/subresource-signing.js (hunk 4 above)
Comment (patricknelson) [inline src/subresource-signing.js:60]: By the way, this was new to me too (learning something). See here for more info: https://kangax.github.io/nfe/#named-expr

I never see this and maybe this is why? When I tried this in Chrome, I couldn't make that available outside of the scope of the function itself. Could have been a Chrome thing but turns out maybe it's spec (didn't see a citation on that claim, however). Mozilla tends to be pretty good at documenting stuff, see here as well: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/function#Named_function_expression

File: README.md (hunk 1 above)
Comment (kojiromike) [inline README.md:28]: Oh, I see what you're saying. Yes, embedding the public key anywhere untrusted undermines the purpose of signing, since the pubkey can be compromised.

File: src/subresource-signing.js (hunk 4 above)
Comment (kojiromike) [inline src/subresource-signing.js:60]: From what I've read here (stipulating that I'm occupied with other things and didn't dig deep) I still don't see a reason to remove the function names.

File: README.md (hunk 1 above)
Comment (scottsb) [inline README.md:28]: I think lines are crossing, actually. While everything said above is true, what I was talking about in my tweet was the _private key_ only: not the public key. My point was that the signing needs to be done separately from the subresource host, since the whole point is to detect a compromise of the subresource host. While perhaps obvious in the broad scheme, this point was significant to me because my thoughts originally went the direction of having the subresource host automatically sign the files. As @kojiromike clarified in response, that's not the intent: https://twitter.com/michaelasmith/status/676520007879143424

File: src/subresource-signing.js (hunk 4 above)
Comment (patricknelson) [inline src/subresource-signing.js:60]: Other than the fact that they technically serve no programmatic function (if you excuse the pun).

File: README.md (hunk 1 above)
Comment (kojiromike) [inline README.md:28]: Right, the distinction between this approach and SRI is that SRI hashes the resources themselves, whereas signing is about trusting the identity of the producer of some resource, regardless of where the resource itself comes from. The host itself doesn't have to do with it. I could download jQuery from 4chan and be sure it's the same jQuery I got from jquery.com. Interestingly, I could see a hierarchical approach working here. If you trust foo's pubkey, and that pubkey signs bar's pubkey, and bar's pubkey signs a bunch of subresources, technically we should allow those resources, even if bar's pubkey (signed by foo) is discovered in an otherwise-untrusted location.

File: src/subresource-signing.js (hunk 4 above)
Comment (scottsb) [inline src/subresource-signing.js:60]: The article @patricknelson posted seems to argue against his own point. The reason for using named functions is for easier debugging in stack traces:

> What it all boils down to is the fact that named function expressions is the only way to get a truly robust stack inspection.

Comment (patricknelson) [inline src/subresource-signing.js:60]: Yeah that's a good point @scottsb. I've gotten bogged down in a pretty intense closure hell when debugging JS exceptions/`Error`'s. Closure? Anonymous function? WHICH ONE?!@! 

Question is -- is that really necessary here? Then again it's not a big deal either way.

Comment (kojiromike) [inline src/subresource-signing.js:60]: The way I see it, the burden of proof of necessity is on the one who made the most recent change ;)

Comment (patricknelson) [inline src/subresource-signing.js:60]: I thought it would be on the one who produced the most programming constructs?! Ok fine...

[DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) + Readability > Stack traces 

Readability since it's an uncommon syntax and confused me at first as well. Also justifying stack traces since at this level since the only item changed here prone to failure is not ... well it's just not something that needs a named function for stack traces if you're attaching a script to the DOM.
